File: dotnet/src/AI.Sentinel.AspNetCore/Remote/RemoteEvaluationClient.cs
using System.Diagnostics;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using AI.Sentinel.AspNetCore.Contract;
using AI.Sentinel.AspNetCore.Internal;
using AI.Sentinel.AspNetCore.Observability;
using AI.Sentinel.AspNetCore.Options;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace AI.Sentinel.AspNetCore.Remote;

/// <summary>
/// Authenticated HTTP client for remote evaluation. No automatic retries.
/// Transport failures yield REMOTE_EVALUATION_FAILURE fail-open responses.
/// </summary>
public sealed class RemoteEvaluationClient : IRemoteEvaluationClient
{
    private readonly HttpClient _httpClient;
    private readonly AiSentinelOptions _options;
    private readonly ISentinelTelemetry _telemetry;
    private readonly ILogger<RemoteEvaluationClient> _logger;

    public RemoteEvaluationClient(
        HttpClient httpClient,
        IOptions<AiSentinelOptions> options,
        ISentinelTelemetry telemetry,
        ILogger<RemoteEvaluationClient> logger)
    {
        _httpClient = httpClient;
        _options = options.Value;
        _telemetry = telemetry;
        _logger = logger;
    }

    public async Task<EvaluationResponse> EvaluateAsync(
        EvaluationRequest request,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        _telemetry.RecordAttempt();
        var stopwatch = Stopwatch.StartNew();
        var outcome = RemoteEvaluationOutcome.UNEXPECTED;

        try
        {
            using var httpRequest = new HttpRequestMessage(HttpMethod.Post, _options.EvaluationPath)
            {
                Content = JsonContent.Create(request, options: SentinelJson.Options)
            };
            httpRequest.Headers.TryAddWithoutValidation(
                RemoteEvaluationConstants.ApiKeyHeader,
                _options.ApiKey);

            using var httpResponse = await _httpClient.SendAsync(
                httpRequest,
                HttpCompletionOption.ResponseContentRead,
                cancellationToken).ConfigureAwait(false);

            if (httpResponse.StatusCode == System.Net.HttpStatusCode.Unauthorized
                || httpResponse.StatusCode == System.Net.HttpStatusCode.Forbidden)
            {
                outcome = RemoteEvaluationOutcome.AUTH_REJECTED;
                return Fail(request.CorrelationId, outcome);
            }

            if (!httpResponse.IsSuccessStatusCode)
            {
                outcome = RemoteEvaluationOutcome.HTTP_ERROR;
                return Fail(request.CorrelationId, outcome);
            }

            var raw = await httpResponse.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            if (string.IsNullOrWhiteSpace(raw))
            {
                outcome = RemoteEvaluationOutcome.MALFORMED_RESPONSE;
                return Fail(request.CorrelationId, outcome);
            }

            EvaluationResponse? response;
            try
            {
                response = JsonSerializer.Deserialize<EvaluationResponse>(raw, SentinelJson.Options);
            }
            catch (JsonException)
            {
                outcome = RemoteEvaluationOutcome.MALFORMED_RESPONSE;
                return Fail(request.CorrelationId, outcome);
            }

            if (response == null)
            {
                outcome = RemoteEvaluationOutcome.MALFORMED_RESPONSE;
                return Fail(request.CorrelationId, outcome);
            }

            try
            {
                EvaluationResponseValidator.Validate(response, request.CorrelationId);
            }
            catch (EvaluationContractException ex)
            {
                var message = ex.Message ?? string.Empty;
                if (message.Contains("contractVersion", StringComparison.Ordinal))
                {
                    outcome = RemoteEvaluationOutcome.VERSION_MISMATCH;
                }
                else if (message.Contains("correlationId", StringComparison.Ordinal))
                {
                    outcome = RemoteEvaluationOutcome.CORRELATION_MISMATCH;
                }
                else
                {
                    outcome = RemoteEvaluationOutcome.MALFORMED_RESPONSE;
                }

                return Fail(request.CorrelationId, outcome);
            }

            outcome = RemoteEvaluationOutcome.SUCCESS;
            _telemetry.RecordSuccess(response.Action!.Value.ToString(), response.EvaluationStatuses);
            return response;
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            outcome = RemoteEvaluationOutcome.TIMEOUT;
            return Fail(request.CorrelationId, outcome);
        }
        catch (TaskCanceledException)
        {
            outcome = RemoteEvaluationOutcome.TIMEOUT;
            return Fail(request.CorrelationId, outcome);
        }
        catch (HttpRequestException)
        {
            outcome = RemoteEvaluationOutcome.CONNECTION_FAILURE;
            return Fail(request.CorrelationId, outcome);
        }
        catch (JsonException)
        {
            outcome = RemoteEvaluationOutcome.SERIALIZATION_FAILURE;
            return Fail(request.CorrelationId, outcome);
        }
        catch (Exception)
        {
            outcome = RemoteEvaluationOutcome.UNEXPECTED;
            return Fail(request.CorrelationId, outcome);
        }
        finally
        {
            stopwatch.Stop();
            _telemetry.RecordOutcome(outcome);
            _telemetry.RecordLatency(stopwatch.Elapsed);
        }
    }

    private EvaluationResponse Fail(string correlationId, RemoteEvaluationOutcome outcome)
    {
        _logger.LogWarning(
            "Remote evaluation failed outcome={Outcome} correlationId={CorrelationId}",
            outcome,
            correlationId);
        _telemetry.RecordFailure(outcome);
        return EvaluationFailureResponses.RemoteFailure(correlationId);
    }
}

File: ai-sentinel-spring-boot-starter/src/main/java/dev/aisentinel/autoconfigure/evaluation/RemoteEvaluationClient.java
package dev.aisentinel.autoconfigure.evaluation;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.aisentinel.core.contract.EvaluationContractException;
import dev.aisentinel.core.contract.EvaluationFailureResponses;
import dev.aisentinel.core.contract.EvaluationRequest;
import dev.aisentinel.core.contract.EvaluationResponse;
import dev.aisentinel.core.contract.EvaluationResponseValidator;
import dev.aisentinel.core.metrics.FailOpenReason;
import dev.aisentinel.core.metrics.RemoteEvaluationOutcome;
import dev.aisentinel.core.metrics.SentinelMetrics;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.MediaType;
import org.springframework.http.client.JdkClientHttpRequestFactory;
import org.springframework.web.client.RestClient;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestClientResponseException;

import java.net.http.HttpClient;
import java.time.Duration;
import java.util.Objects;

/**
 * Authenticated HTTP client for remote evaluation.
 * <p>
 * No automatic retries (evaluation may mutate server baseline/quarantine state).
 * Transport failures yield {@link EvaluationFailureResponses#remoteFailure(String)} — not fabricated risk.
 */
public final class RemoteEvaluationClient {

    private static final Logger log = LoggerFactory.getLogger(RemoteEvaluationClient.class);

    private final RestClient restClient;
    private final ObjectMapper objectMapper;
    private final SentinelMetrics metrics;
    private final String apiKey;
    private final String evaluationPath;

    public RemoteEvaluationClient(String baseUrl,
                                  String evaluationPath,
                                  String apiKey,
                                  Duration connectTimeout,
                                  Duration readTimeout,
                                  ObjectMapper objectMapper,
                                  SentinelMetrics metrics) {
        Objects.requireNonNull(baseUrl, "baseUrl");
        Objects.requireNonNull(apiKey, "apiKey");
        this.apiKey = apiKey;
        this.evaluationPath = normalizePath(evaluationPath);
        this.objectMapper = Objects.requireNonNull(objectMapper, "objectMapper");
        this.metrics = metrics != null ? metrics : SentinelMetrics.NOOP;

        Duration connect = connectTimeout != null ? connectTimeout : Duration.ofMillis(500);
        Duration read = readTimeout != null ? readTimeout : Duration.ofSeconds(2);
        HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(connect)
            .build();
        JdkClientHttpRequestFactory requestFactory = new JdkClientHttpRequestFactory(httpClient);
        requestFactory.setReadTimeout(read);

        String root = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.restClient = RestClient.builder()
            .baseUrl(root)
            .requestFactory(requestFactory)
            .build();
    }

    /**
     * POST EvaluationRequest once (no retry). On any transport/contract failure returns remote-failure response.
     */
    public EvaluationResponse evaluate(EvaluationRequest request) {
        Objects.requireNonNull(request, "request");
        metrics.recordRemoteEvaluationAttempt();
        long start = System.nanoTime();
        RemoteEvaluationOutcome outcome = RemoteEvaluationOutcome.UNEXPECTED;
        try {
            byte[] body = objectMapper.writeValueAsBytes(request);
            String raw = restClient.post()
                .uri(evaluationPath)
                .contentType(MediaType.APPLICATION_JSON)
                .header(RemoteEvaluationConstants.API_KEY_HEADER, apiKey)
                .body(body)
                .retrieve()
                .body(String.class);

            if (raw == null || raw.isBlank()) {
                outcome = RemoteEvaluationOutcome.MALFORMED_RESPONSE;
                return fail(request.correlationId(), outcome);
            }
            EvaluationResponse response;
            try {
                response = objectMapper.readValue(raw, EvaluationResponse.class);
            } catch (Exception parseEx) {
                outcome = RemoteEvaluationOutcome.MALFORMED_RESPONSE;
                return fail(request.correlationId(), outcome);
            }
            try {
                EvaluationResponseValidator.validate(response, request.correlationId());
            } catch (EvaluationContractException vex) {
                String msg = vex.getMessage() != null ? vex.getMessage() : "";
                if (msg.contains("contractVersion")) {
                    outcome = RemoteEvaluationOutcome.VERSION_MISMATCH;
                } else if (msg.contains("correlationId")) {
                    outcome = RemoteEvaluationOutcome.CORRELATION_MISMATCH;
                } else {
                    outcome = RemoteEvaluationOutcome.MALFORMED_RESPONSE;
                }
                return fail(request.correlationId(), outcome);
            }
            outcome = RemoteEvaluationOutcome.SUCCESS;
            metrics.recordEvaluationStatuses(
                response.evaluationStatuses().stream()
                    .map(name -> {
                        try {
                            return dev.aisentinel.core.decision.EvaluationStatus.valueOf(name);
                        } catch (IllegalArgumentException ignored) {
                            return null;
                        }
                    })
                    .filter(Objects::nonNull)
                    .toList());
            return response;
        } catch (RestClientResponseException httpEx) {
            int status = httpEx.getStatusCode().value();
            if (status == 401 || status == 403) {
                outcome = RemoteEvaluationOutcome.AUTH_REJECTED;
            } else {
                outcome = RemoteEvaluationOutcome.HTTP_ERROR;
            }
            return fail(request.correlationId(), outcome);
        } catch (RestClientException restEx) {
            String name = restEx.getClass().getSimpleName().toLowerCase();
            String msg = restEx.getMessage() != null ? restEx.getMessage().toLowerCase() : "";
            if (name.contains("timeout") || msg.contains("timed out") || msg.contains("timeout")) {
                outcome = RemoteEvaluationOutcome.TIMEOUT;
            } else {
                outcome = RemoteEvaluationOutcome.CONNECTION_FAILURE;
            }
            return fail(request.correlationId(), outcome);
        } catch (Exception ex) {
            String name = ex.getClass().getSimpleName().toLowerCase();
            if (name.contains("json") || name.contains("jackson")) {
                outcome = RemoteEvaluationOutcome.SERIALIZATION_FAILURE;
            } else {
                outcome = RemoteEvaluationOutcome.UNEXPECTED;
            }
            return fail(request.correlationId(), outcome);
        } finally {
            metrics.recordRemoteEvaluationLatencyNanos(System.nanoTime() - start);
            metrics.recordRemoteEvaluationOutcome(outcome.name());
        }
    }

    private EvaluationResponse fail(String correlationId, RemoteEvaluationOutcome outcome) {
        log.warn("Remote evaluation failed outcome={} correlationId={}", outcome, correlationId);
        metrics.recordFailOpen(FailOpenReason.REMOTE_EVALUATION_FAILURE);
        return EvaluationFailureResponses.remoteFailure(correlationId);
    }

    private static String normalizePath(String path) {
        if (path == null || path.isBlank()) {
            return "/ai-sentinel/v1/evaluation";
        }
        return path.startsWith("/") ? path : "/" + path;
    }
}
